Switch analysis tasks to be classes

xylar opened the pull request:
Adds a base class, `AnalysisTask` and template file from which to create new analysis tasks.

Updates analysis tasks to match new template

Moves functions `check_path_exists` and `make_directories` to `shared.io.utility`

Updates `run_analysis.py` to create a list of analyses. Gets task and component names and a list of categories from analysis tasks. Calls `check` functions to make sure config is correct and files are present. Later, calls `run` on each analysis task.

xylar commented:
This PR is intended as an implementation of the design doc in#127.

It aims to address issues #11 (providing user-friendly examples).

It is intended, in part, to be a helpful reorganization step prior to #112 (task parallelism).

xylar commented:
### Testing

I have analyzed the `GMPAS-QU240` test case with all tasks on my laptop.  The results are bit-for-bit with those from the `horiz_interp` branch (#123) with the same parameters.

I have verified that, when `check` or `run` for a given task fails, it prints an error including the stack trace but then continues with the next task without exiting analysis completely.

More testing is obviously needed and I will update when I have time to perform them.

xylar commented:
@milenaveneziani, this would be the next one after #131.  But I want to give @vanroekel and @pwolfram a chance to look it over and review it if they want to after #131 gets merged and I rebase this.

xylar commented:
@milenaveneziani, @vanroekel and @pwolfram, this PR has been rebase onto `develop` and should be ready for a review.  Also, please take a look at the design document, #127.

xylar commented:
Adding back the "don't merge" flag until we have updated `ACMEPreAndPostProcessingScripts`.

xylar commented:
@milenaveneziani, @vanroekel and @pwolfram, are you okay with me changing `modelvsobs`, `regridded` and `regriddedHorizontal` (in file names, function names, config sections and elements of the `generate` option) to either `climatologyMap` (for everything but file names) or `climatology_map` (file names)?  I feel this describes more accurately what we mean.  

I would like to make this change in this PR (and I actually already did some of it in my last commit).

milenaveneziani commented:
@xylar: when you have a chance, could you please rebase all the remaining PR's so it'll be easier to review them? I still think we should merge them after we update the ACME submodule, but perhaps we can review them earlier. Thanks.

xylar commented:
@milenaveneziani, I'm going to have to do some pretty significant updating to this branch because of your recent changes so I might not get around to rebasing for a little bit.  Please hold off on reviewing this (and also parallel tasks) for now.  I'll let you know when it's ready for your attention.

I do think we want to include this and parallel tasks in the "official release" of QUIDA or whatever it's called.

milenaveneziani commented:
> I do think we want to include this and parallel tasks in the "official release" of QUIDA or whatever it's called.

Definitely. But for me it would be easier to do it in 2 steps. First what we have now, plus the new xarray and Nino3.4 changes (if possible). Then the parallel tasks capability.

xylar commented:
Okay, sounds good.  The 2-step process was what I was thinking, too.

xylar commented:
@milenaveneziani, @pwolfram, @vanroekel and @mark-petersen, I would like to make this PR a priority after v0.3 is created.  The reason is that it is pretty fundamental to the way new analysis tasks are created and the more new analysis that gets added, the bigger this PR will have to become.

I have broken the PR into a bunch of commits, one introducing the base class `AnalysisTask` (and the template from which to create analysis) and then one commit each for the subsequent conversion of each analysis task to a class.  Major changes to `run_analysis.py` are also included with the introduction of the first analysis task (ocean climatology maps).  My hope is that this will make the changes easier to follow and will make any future debugging a bit simpler.

xylar commented:
Clearly, I have some work to do on unit tests.  I'll look into that now.

xylar commented:
@milenaveneziani, I just rebased this branch after the latest merges and cleaned up a conflict. 

I would appreciate it if you could take a look at this sometime in the next week or 2.  As I said, it's a pretty big change, which means that it would be best to have it merged before we add anymore analysis tasks.

I'll be pretty busy in the next week, and then not available at all between the 16th and the 25th,  but I'll try to do what i can to address any concerns that come up.

milenaveneziani commented:
@xylar: I am at a meeting for the first 3 days of this week. Can anybody else start taking a look in the meantime? I think it would probably be good to have a short phone call (Th?), after I go over the main changes for adding new analysis tasks.

xylar commented:
@milenaveneziani, yep, let's talk by phone when you have time.  I don't know if anyone else has any time to look this over.

milenaveneziani commented:
@xylar: I know we agreed on the phone that I could implement a simple task, or even better one that I need to work on anyway, to test this workflow out. But I don't think I have the time to do this and review the PR by the end of this week. So, here is my suggestion: 1) I will still review `analysis_task_template` and provide feedback on additional documentation where I think it's necessary, and 2) what I think it could be useful to add, at the same time, is a how-to guide. This guide would essentially be a list of steps that an MPAS-Analysis developer would have to follow to add a new task. I can be a little more specific as I go through the code, within the next day or so.

Let me know what you think.

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,133 @@
+'''
+This is an example analysis task to be used as a template for new tasks.
+It should be copied into one of the component folders (`ocean`, `sea_ice`,
+`land_ice`, etc.) and modified as needed.
```
Add something along these lines:
`Specific parts or variable names that need to be modified by the developer are explicitly indicated. All the others are to be left unchanged.`

xylar replied:
Added

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,133 @@
+# dots)
+from ..shared.analysis_task import AnalysisTask
+
+
```
Add comment: `Change MyTaskName below`.
I would also rename `MyTask` with `MyTaskName`.

xylar replied:
I disagree with the change from `MyTask` to `MyTaskName` because the ending `Name` is appropriate for a string containing the task name but not for a class that implements a task.  So the class really should be thought of as *being* the task (as opposed to just having the same name as the task).

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,133 @@
+
+    def setup_and_check(self):  # {{{
+        '''
+        Perform steps to set up the analysis and check for errors in the setup.
```
Add a bit more explanation about what kind of checks we should do here.

xylar replied:
I added more explanation as a comment (and a bit more to the docstring, too).

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,133 @@
+
+    def run(self):  # {{{
+        '''
+        Runs the analysis task.
```
Perhaps change to `Main script that defines and runs the analysis task`

xylar replied:
I modified the docstring similarly to your suggestion.

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,133 @@
+        <List of authors>
+        '''
+
+        # here is where the main 'meat' of the analysis task goes
```
Change to: 
```
Add the main 'meat' of the analysis task below

Here is an example of a call to a local helper function:
```

xylar replied:
Changed as you suggest.

milenaveneziani commented:
It was easier than I thought.
Here are the steps that I *think* a developer would have to follow, let me know if I have missed anything. I also added some information that I think it would be good to include in this How-to guide:
```
Steps for adding new MPAS-Analysis task:
1) create new task by copying mpas_analysis/analysis_task_template to appropriate folder and
modifying it as necessary. [Maybe we can also add that one can look at mpas_analysis/shared/analysis_task.py for additional guidance?]
2) modify config.default (and any machine-specific config file if necessary)
3) import new analysis task in mpas_analysis/componentName/_init_.py
4) [Milena's comment: I don't think mpas_analysis/shared/analysis_task.py needs to be changed, right?]
5) if necessary, update variable maps in mpas_analysis/shared/variable_namelist_stream_maps/*_maps.py
6) add new analysis task to run_analysis. [Here, it would also be good to describe how run_analysis is structured now, and where exactly the analysis task in executed (at analysisTask.run())]
```

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,217 @@
+2. modify config.default (and possibly any machine-specific config files in
+   configs/<machine>)
+3. import new analysis task in mpas_analysis/<component>/__init__.py
+4. note, no changes need to be made to mpas_analysis/shared/analysis_task.py
```
Thanks for adding these steps here as well, @xylar.
Maybe I would move 4. to 2. (right after 'Take a look..', so that it is clear that analysis_task can be referenced to, but doesn't have to be modified).

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,217 @@
+   mpas_analysis/shared/variable_namelist_stream_maps/<component>_maps.py
+6. add new analysis task to run_analysis.py under build_analysis_list:
+      analyses.append(<component>.MyTask(config, myArg='argValue'))
+   will add a new object of the MyTask class to a list of analysis tasks
```
small addition above: `this will add a new object..`

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,217 @@
+   created in build_analysis_list.  Later on in run_analysis, it will first
+   go through the list to make sure each task needs to be generated
+   (by calling check_generate, which is defined in AnalysisTask), then, will
+   call setup_and_check on each task (to make sure the appropriate AM is on
```
I still think we can reduce complexity by removing `setup_and_check`, unless I am the only one advocating for this :)

xylar replied:
This is substantially more work than the other changes you requested.  I think it's conceptually useful to separate checking whether the setup is correct from running the task.  If we are going to request that each task do this, I don't see a big advantage in this being a helper function that each task runs at the beginning of `run` rather than being a separate function that they are required to implement.

xylar replied:
Perhaps you could explain your concern about the complexity you feel `setup_and_check` adds to the process so I could better understand why this is an added burden.

xylar commented:
I just made a change so `AnalysisTask` gets constructed with 3 arguments: `taskName`, `componentName` and `tags`, which it stores.  This will ensure that no task gets created without these items being supplied and stored.

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,217 @@
+class MyTask(AnalysisTask):  # {{{
+    '''
+    <Briefly describe the analysis task here.  Just a 1-2 sentence description
+    of what the task does should do.>
```
small typo above

xylar replied:
I don't see a typo.  Perhaps "should do" is confusing so I will change it to "should be sufficient".

milenaveneziani replied:
yes, sorry, I read it as either 'does' or 'should do'.

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,217 @@
+    # create a new object from your class.  Typically, it will just store a few
+    # useful pieces of information like the name of the task, the component and
+    # maybe some tags describing the task.
+    # Keep self and config arguments.  Either modify myArg to be a useful
```
Is it appropriate to mention what `self` is and why is `config` necessary? A basic explanation similar to our phone conversation.

xylar replied:
Okay, I added one.

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,217 @@
+        # time series, indices and/or climatologies should include
+        # 'timeSeries', 'index' and/or 'climatology' tags, as these are helpful
+        # used to compute and update the start and end dates of each of these
+        # analysis tasks as part of setting up the task.
```
very nice explanation above, thanks.
Similar to the comment about `self`, would it be OK to mention briefly what `super` does?

xylar replied:
added

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,227 @@
+        # likely include things like the name of a field, region, month,
+        # season, etc. to be analyzed so that the same subclass of AnalysisTask
+        # can perform several different tasks (potentially in parallel)
+        # Modify or remove the folowing:
```
At this point in the template, it is not clear to me what I would have to add as `MyArg` (if anything). When you say `later on` above, I suppose you mean in `setup_and_check` and in `run`? And how do I know what arguments are already there (from analysis_task?), so that I can add the ones that I am missing for the specific task?

xylar replied:
So I don't really know how to address this better than what I already have done here.  If you had a task that was capable of analyzing multiple fields, the field name should be an argument.  I would advocate breaking existing tasks that analyze different seasons into separate tasks, one for each season (or at least having the option to do this), in which case the name of the season ('DJF', 'ANN') would be an argument you would use to construct the task.  But there's no way to know what arguments a task might need in advance.  This is simply part of programming a new task.  I'm just trying to show how you would add arguments containing information you *knew* you needed to pass to the task and couldn't get from config options directly.  We're not currently using this capability, but an example would have been the field names ('sst', 'sss' and 'mld') passed to ocean_modelvsobs.

milenaveneziani replied:
ok, I guess the best way is to just look at an existent analysis task.

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,227 @@
+        # }}}
+
+    # this function will be called to figure out if the analysis task should
+    # run.  It shoul check if the input arguments to the task are supported,
```
small typo (should)

milenaveneziani commented on mpas_analysis/analysis_task_template.py:
```diff
@@ -0,0 +1,227 @@
+            raise ValueError('MyTask must be constructed with argument myArg '
+                             'having one of the values\n'
+                             '{}.'.format(possibleArgs))
+
```
Thanks: the examples below are exactly what I was going to suggest. We can leave setup_and_check as is.

milenaveneziani commented:
@xylar: I am happy with your changes to this PR. I feel like things are much more documented now, and we can always revisit anything once one of us rather than you makes a new analysis task. Thanks for your patience. I will wait one more day to give @vanroekel and @pwolfram a chance to add any comment, and then merge this by the end of the weekend.
As a heads up: the idea is to have this PR and #129 in v0.4, so that we can update A-Prime in the next couple of weeks. Does this seem like a good plan to you?

milenaveneziani commented:
sorry, I just realized that the configs/machine files have not been changed to reflect the changes in config.default (I am mainly thinking about the different task names). I know it's boring: maybe I could help with this?

xylar commented:
> sorry, I just realized that the configs/machine files have not been changed to reflect the changes in config.default (I am mainly thinking about the different task names). I know it's boring: maybe I could help with this?

Oh, I guess I hadn't realized this was an issue.  I'll take a look when I can.

milenaveneziani commented:
I added the changes to the machine config files, @xylar. It was boring but simple.
I also ran a test on edison on the beta1_05 run and all worked well.

xylar commented:
Thanks, @milenaveneziani, for adding those tedious but very necessary changes to the machine-specific config files.

milenaveneziani commented:
A test on the GMPAS-IAF_oRRS18to6v3_CIME53 case on titan passed as expected.
